hazelcast-client/src/main/java/com/hazelcast/client/connection/nio/ClientConnectionManagerImpl.java

```diff
@@ -377,6 +378,9 @@ private Connection getConnection(Address target, boolean asOwner, boolean acquir
     }
 
     private void checkAllowed(Address target, boolean asOwner, boolean acquiresResources) throws IOException {
+        if (!alive) {
```

Comment:
`HazelcastClientNotActiveException` does not extend `HazelcastException`. This is a behaviour change then. Is it possible some user was catching a `HazelcastException?`

Reply:
HazelcastException is retryable in client invocation system. When hazelcast exception is thrown from here, invocation system catches that throws as ClientNotActiveException to user. 
https://github.com/hazelcast/hazelcast/blob/master/hazelcast-client/src/main/java/com/hazelcast/client/spi/impl/ClientInvocation.java#L209
But there is an unncessary wrapping there. It results as a ClientNotActiveException which has HazelcastException("client is shutting down") as cause.

There are a couple, usages of connection manager directly like txns which should handle HazelcastException themselves. In that case, we also convert it to ClientNotActiveException

So in almost all of the cases, user get ClientNotActiveException. Even if there were edge cases that user gets HazelcastException before, I would like to consider them as bug and fix it, rather then thinking this as behaviour change. 
How does that sound ?

Reply:
I can admit I'm a bit lost here. It's your call. I'm just thinking if this could affect some user code like:
```
try {
  // use hazelcast
} catch (HazelcastException expected) {
  // all is good, I expected this!
}
```

Reply:
This is also my worry.

Reply:
Well as mentioned above, if we are not throwing `ClientNotActiveException` where we supposed to throw it is a bug. 
And depending on when the clients shutdown we could be throwing `ClientNotActiveException` or `HazelcastException`. That means, user should already prepared for  `ClientNotActiveException`. This fix is to correct one more place that we are throwing `HazelcastException` instead of `ClientNotActiveException`. 

That is why, I see no problem with backward compatibility.

Reply:
IllegalStateException can't derive from HazelcastException, because it is from java.lang library. It is not our class.